File: leetcode/python/242_valid_anagram.py

from collections import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        counter = Counter(t)
        for c in s:
            if c not in counter or counter[c] == 0:
                return False
            counter[c] -= 1
        return True

s = Solution()
print(s.isAnagram('rat', 'tart'))
print(s.isAnagram(s = "anagram", t = "nagaram"))
print(s.isAnagram(s = "rat", t = "car"))

File: leetcode/rust/src/242_valid_anagram.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() { return false; }

        let mut counter = HashMap::new();
        for c in s.chars() {
            let count = counter.entry(c).or_insert(0);
            *count += 1;
        }

        for c in t.chars() {
            if counter.contains_key(&c) && *counter.get(&c).unwrap() > 0 {
                let count = counter.entry(c).or_insert(0);
                *count -= 1;
            } else {
                return false;
            }
        }
        return true;
    }
}